In [5]:
import pandas as pd

In [7]:
import numpy as np

In [11]:
df = pd.read_csv('Ecommerce_Analysis.csv')

In [13]:
df = pd.read_csv('Ecommerce_analysis.csv',nrows =1000001)

In [14]:
df.head()

,user_id,product_id,category,subcategory,brand,price,discount,final_price,rating,review_count,stock,seller_id,seller_rating,purchase_date,shipping_time_days,location,device,payment_method,is_returned,delivery_status
0,U356787,P39256,Electronics,Mobile,H&M,33090.51,10.46,29629.31,4.9,43,71,S2679,4.3,04-03-2025,6,Bangalore,Tablet,UPI,True,Returned
1,U198246,P38657,Sports,Outdoor,Samsung,9368.97,5.32,8870.73,3.9,13,119,S9279,4.9,12-12-2025,1,Delhi,Tablet,UPI,True,Returned
2,U539898,P38893,Sports,Fitness,H&M,14756.85,26.00,10919.43,4.2,46,229,S5557,4.9,25-04-2024,1,Chennai,Mobile App,Credit Card,True,Returned
3,U325772,P54118,Beauty,Makeup,Sony,668.83,24.81,502.91,3.8,6,52,S2519,3.1,18-12-2025,2,Mumbai,Web,UPI,False,In Transit
4,U865179,P70217,Home,Decor,Sony,10881.29,48.63,5589.58,4.0,48,274,S3045,2.6,16-05-2024,5,Delhi,Web,UPI,False,Delayed


In [17]:
df.shape

(1000000, 20)

In [18]:
df.head()

,user_id,product_id,category,subcategory,brand,price,discount,final_price,rating,review_count,stock,seller_id,seller_rating,purchase_date,shipping_time_days,location,device,payment_method,is_returned,delivery_status
0,U356787,P39256,Electronics,Mobile,H&M,33090.51,10.46,29629.31,4.9,43,71,S2679,4.3,04-03-2025,6,Bangalore,Tablet,UPI,True,Returned
1,U198246,P38657,Sports,Outdoor,Samsung,9368.97,5.32,8870.73,3.9,13,119,S9279,4.9,12-12-2025,1,Delhi,Tablet,UPI,True,Returned
2,U539898,P38893,Sports,Fitness,H&M,14756.85,26.00,10919.43,4.2,46,229,S5557,4.9,25-04-2024,1,Chennai,Mobile App,Credit Card,True,Returned
3,U325772,P54118,Beauty,Makeup,Sony,668.83,24.81,502.91,3.8,6,52,S2519,3.1,18-12-2025,2,Mumbai,Web,UPI,False,In Transit
4,U865179,P70217,Home,Decor,Sony,10881.29,48.63,5589.58,4.0,48,274,S3045,2.6,16-05-2024,5,Delhi,Web,UPI,False,Delayed


In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 20 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   user_id             1000000 non-null  object 
 1   product_id          1000000 non-null  object 
 2   category            1000000 non-null  object 
 3   subcategory         1000000 non-null  object 
 4   brand               1000000 non-null  object 
 5   price               1000000 non-null  float64
 6   discount            1000000 non-null  float64
 7   final_price         1000000 non-null  float64
 8   rating              1000000 non-null  float64
 9   review_count        1000000 non-null  int64  
 10  stock               1000000 non-null  int64  
 11  seller_id           1000000 non-null  object 
 12  seller_rating       1000000 non-null  float64
 13  purchase_date       1000000 non-null  object 
 14  shipping_time_days  1000000 non-null  int64  
 15  location        

In [22]:
df['purchase_date'] = pd.to_datetime(df['purchase_date'],dayfirst =True)

In [25]:
df.isnull().sum()
df['purchase_date'].head()

0   2025-03-04
1   2025-12-12
2   2024-04-25
3   2025-12-18
4   2024-05-16
Name: purchase_date, dtype: datetime64[ns]

In [26]:
df.isnull().sum()

user_id               0
product_id            0
category              0
subcategory           0
brand                 0
price                 0
discount              0
final_price           0
rating                0
review_count          0
stock                 0
seller_id             0
seller_rating         0
purchase_date         0
shipping_time_days    0
location              0
device                0
payment_method        0
is_returned           0
delivery_status       0
dtype: int64

In [27]:
df.describe()

,price,discount,final_price,rating,review_count,stock,seller_rating,purchase_date,shipping_time_days
count,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000,1000000.000000
mean,13224.119093,28.996161,9938.876985,3.927226,32.626704,250.023345,3.750030,2025-03-30 23:32:00.384000256,3.166389
min,200.030000,5.000000,101.460000,1.500000,0.000000,0.000000,2.500000,2024-03-31 00:00:00,1.000000
25%,2700.147500,16.910000,1705.930000,3.600000,10.000000,125.000000,3.100000,2024-09-29 00:00:00,2.000000
50%,4971.830000,28.070000,3790.795000,3.900000,20.000000,250.000000,3.800000,2025-03-31 00:00:00,3.000000
75%,14134.482500,39.240000,10384.440000,4.300000,39.000000,375.000000,4.400000,2025-09-30 00:00:00,5.000000
max,79999.700000,70.000000,75831.110000,5.000000,3408.000000,500.000000,5.000000,2026-03-31 00:00:00,6.000000
std,18079.626921,14.681077,14271.885557,0.519443,43.382201,144.567038,0.722092,NaN,1.674178


In [35]:
df['expected_final'] = df['price'] * (1-df['discount']/100)
df['price_diff'] = df['final_price']-df['expected_final']
df['price_diff'].describe()

count    1000000.000000
mean           0.000057
std            0.645587
min           -3.986548
25%           -0.119672
50%            0.000216
75%            0.120272
max            3.986772
Name: price_diff, dtype: float64

In [36]:
df =df.drop(columns=['expected_final', 'price_diff'])

In [39]:
df.groupby('category')['brand'].value_counts()

category     brand  
Beauty       Puma       16842
             LG         16703
             Lenovo     16694
             HP         16679
             Apple      16653
             Sony       16645
             Boat       16616
             Samsung    16602
             Nike       16546
             Zara       16520
             Adidas     16481
             H&M        16346
Clothing     Samsung    16799
             HP         16753
             Boat       16748
             LG         16738
             Apple      16718
             Zara       16717
             Nike       16643
             Puma       16621
             Sony       16586
             Lenovo     16567
             Adidas     16486
             H&M        16448
Electronics  Samsung    16801
             Lenovo     16790
             Sony       16748
             Boat       16705
             Nike       16693
             LG         16663
             Adidas     16661
             HP         16642
             Apple 

In [41]:
df.groupby('brand')['category'].nunique()

brand
Adidas     5
Apple      5
Boat       5
H&M        5
HP         5
LG         5
Lenovo     5
Nike       5
Puma       5
Samsung    5
Sony       5
Zara       5
Name: category, dtype: int64

In [43]:
df.nunique()

user_id               603815
product_id             89999
category                   5
subcategory               16
brand                     12
price                 765505
discount                6501
final_price           713388
rating                    36
review_count             818
stock                    501
seller_id               9000
seller_rating             26
purchase_date            731
shipping_time_days         6
location                   5
device                     3
payment_method             4
is_returned                2
delivery_status            4
dtype: int64

In [45]:
df.groupby('brand')['category'].nunique()

brand
Adidas     5
Apple      5
Boat       5
H&M        5
HP         5
LG         5
Lenovo     5
Nike       5
Puma       5
Samsung    5
Sony       5
Zara       5
Name: category, dtype: int64

In [47]:
df['location'].unique()

array(['Bangalore', 'Delhi', 'Chennai', 'Mumbai', 'Hyderabad'],
      dtype=object)

In [48]:
df['device'].unique()

array(['Tablet', 'Mobile App', 'Web'], dtype=object)

In [50]:
df['payment_method'].unique()

array(['UPI', 'Credit Card', 'Debit Card', 'Cash on Delivery'],
      dtype=object)

In [51]:
df['delivery_status'].unique()

array(['Returned', 'In Transit', 'Delayed', 'Delivered'], dtype=object)

In [52]:
# Category highest return rate

In [54]:
.groupby('category')['is_returned'].mean().sort_values(ascending= False)

category
Beauty         0.117561
Clothing       0.117363
Sports         0.116460
Home           0.115841
Electronics    0.112734
Name: is_returned, dtype: float64

In [56]:
# Most Delivery Delays Location

In [55]:
df[df['delivery_status'] == 'Delayed']['location'].value_counts()

location
Delhi        59839
Mumbai       59554
Bangalore    58827
Chennai      58506
Hyderabad    58257
Name: count, dtype: int64

In [58]:
df.groupby('location')['delivery_status'].apply(lambda x:(x=='Delayed').mean()).sort_values(ascending=False)

location
Delhi        0.298227
Mumbai       0.297958
Bangalore    0.293738
Chennai      0.292673
Hyderabad    0.292301
Name: delivery_status, dtype: float64

In [57]:
# Seller Ratings with returns

In [59]:
df.groupby('is_returned')['seller_rating'].mean()

is_returned
False    3.749669
True     3.752778
Name: seller_rating, dtype: float64

In [60]:
df.groupby('is_returned')['shipping_time_days'].mean()

is_returned
False    3.134466
True     3.409690
Name: shipping_time_days, dtype: float64

In [62]:
df.groupby('payment_method')['is_returned'].mean().sort_values(ascending=False)

payment_method
Credit Card         0.116529
Debit Card          0.116176
UPI                 0.115835
Cash on Delivery    0.115421
Name: is_returned, dtype: float64

# E-commerce Returns & Delivery Analysis (1M+ Records)

**Dataset:** 1,000,000 e-commerce transaction records across 20 fields (category, brand, price, seller info, delivery status, returns, etc.)

**Goal:** Identify which factors, if any, are associated with order returns and delivery delays.

**Tools:** Python (pandas) for cleaning and analysis.

## Step 1: Data Profiling

Checked shape, data types, nulls, and summary statistics before any cleaning.

## Step 2: Data Quality Findings

- 1,000,000 rows, 20 columns, **zero missing values** across all fields
- `purchase_date` was stored as text; converted to datetime
- Numeric fields (price, discount, final_price, rating, seller_rating) all fall within logical ranges — no negative prices, no out-of-range ratings
- `final_price` was verified against `price` and `discount`; differences were negligible (rounding only)
- **Brand and category fields are independently randomized** — every one of the 12 brands appears across all 5 categories, confirming synthetic data generation rather than real product-brand relationships. Brand-category pairs are therefore not treated as meaningful signal in this analysis.
- Categorical fields (location, device, payment_method, delivery_status) are consistently formatted with no typos or casing issues

**Conclusion:** minimal cleaning was required beyond the datetime conversion.

## Step 3: Business Questions

1. Which category has the highest return rate?
2. Which location has the highest delivery delay rate?
3. Does seller rating correlate with returns?
4. Does shipping time correlate with returns?
5. Does payment method correlate with returns?

**Finding:** Return rates are nearly identical across categories (11.3%–11.8%). No category shows a meaningfully elevated return risk.

**Finding:** Delay rates are nearly identical across all 5 cities (29.2%–29.8%). Location has no meaningful effect on delivery delays.

**Finding:** Average seller rating is virtually identical for returned vs. non-returned orders (3.75 vs 3.75). Seller rating does not predict returns.

**Finding:** Orders that were returned had a noticeably longer average shipping time (3.41 days) compared to orders that were not returned (3.13 days) — roughly a 9% relative difference. This is the strongest signal found in the analysis.

**Finding:** Return rates are nearly identical across all payment methods (11.5%–11.7%). Payment method has no meaningful effect on returns.

## Final Conclusion

Across five variables tested — category, location, seller rating, payment method, and shipping time — only **shipping time** showed a meaningful relationship with returns. Orders that were eventually returned had ~9% longer average shipping times than orders that were not returned. All other variables tested showed flat, statistically negligible differences.

**Business takeaway:** Returns in this dataset appear to be driven primarily by delivery experience (shipping delays) rather than product category, seller quality, customer location, or payment method. This suggests operational/logistics improvements (faster, more reliable shipping) may have more impact on reducing returns than seller vetting or category-specific interventions.